BatchNorm2D

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61310
* #61265
* #61239
* #61238
* #61188
* #61150
* #61093
* **#61012**

Differential Revision: [D29562337](https://our.internmc.facebook.com/intern/diff/D29562337)

migeed-z committed:
ResNet50

[ghstack-poisoned]

migeed-z committed:
Update on "ResNet50"




[ghstack-poisoned]

migeed-z committed:
Update on "ResNet50"




[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"




[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"




[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"




[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"




[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"




[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"




[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"




[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29562337).

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -184,18 +225,25 @@ def type_check_node(self, n: Node):
+        if n.op == 'call_module':
+            op_type = getattr(self.traced, str(n.target))
+            if type(op_type) in _INFERENCE_RULES:
+                return _INFERENCE_RULES[type(op_type)](n, op_type)
```
We should document somewhere what the schema for inference rules should be since now it's not just the `Node`. For example, without reading this code, I don't know what the `op_type` argument is supposed to be. It seems like it's actually the module instance rather than the type?

migeed-z replied:
Yeah it's the module in this case. I'll rename it.

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -4,10 +4,30 @@
 _INFERENCE_RULES: Dict[Target, Callable] = {}
 
 
+def apply_matching(t, n):
```
Maybe name this something like `expand_to_tensor_dim`? `apply_matching` isn't very descriptive IMO

migeed-z replied:
matching is the name for this operation in gradual types but I can rename it!

jamesr66a reviewed (commented):
Looking good!

jamesr66a commented on test/fx/test_gradual_type.py:
```diff
@@ -235,6 +236,100 @@ def forward(self, x: TensorType((1, 2, 3, 5))):
+        traced = GraphModule(ast_rewriter.root, graph, "gm")
+        tc = GraphTypeChecker({}, traced)
+        with self.assertRaises(TypeError):
+            tc.type_check()
```
When I look at the error message, looks like we lost `_DynType.__repr__`:

```
TypeError: Cannot apply BatchNorm2d(1, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True) with input type TensorType[(<torch.fx.tensor_type._DynType object at 0x7f5537046940>, <torch.fx.tensor_type._DynType object at 0x7f5537046940>, <torch.fx.tensor_type._DynType object at 0x7f5537046940>, <torch.fx.tensor_type._DynType object at 0x7f5537046940>)] and existing type TensorType[(2, 2, <torch.fx.tensor_type._DynType object at 0x7f5537046940>, 4)] on bn1
```

Can you add that?

migeed-z committed:
Update on "BatchNorm2D"


Differential Revision: [D29562337](https://our.internmc.facebook.com/intern/diff/D29562337)

[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"


Differential Revision: [D29562337](https://our.internmc.facebook.com/intern/diff/D29562337)

[ghstack-poisoned]

migeed-z committed:
Update on "BatchNorm2D"


Differential Revision: [D29562337](https://our.internmc.facebook.com/intern/diff/D29562337)

[ghstack-poisoned]

jamesr66a reviewed (approved):
LGTM

migeed-z committed:
Update on "BatchNorm2D"


Differential Revision: [D29562337](https://our.internmc.facebook.com/intern/diff/D29562337)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29562337).
